### PR TITLE
Cache cast function in set during execution

### LINKED ISSUE #54239: Set::execute excessively builds internal CAST functions
image

====

Building functions require parsing, so it can be slow. Probably `Set` should cache the `CAST` corresponding functions for each column.

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Cache cast function in set during execution to improve the performance of function `IN` when set element type doesn't exactly match column type.

Avoid creating same cast function each time calling `Set::execute`. Close #54239

Performance test:

```sql
CREATE TABLE iso_3166_1_alpha_2
(
    `c` Enum8('LI' = -128, 'LT' = -127, 'LU' = -126, 'MO' = -125, 'MK' = -124, 'MG' = -123, 'MW' = -122, 'MY' = -121, 'MV' = -120, 'ML' = -119, 'MT' = -118, 'MH' = -117, 'MQ' = -116, 'MR' = -115, 'MU' = -114, 'YT' = -113, 'MX' = -112, 'FM' = -111, 'MD' = -110, 'MC' = -109, 'MN' = -108, 'ME' = -107, 'MS' = -106, 'MA' = -105, 'MZ' = -104, 'MM' = -103, 'NA' = -102, 'NR' = -101, 'NP' = -100, 'NL' = -99, 'NC' = -98, 'NZ' = -97, 'NI' = -96, 'NE' = -95, 'NG' = -94, 'NU' = -93, 'NF' = -92, 'MP' = -91, 'NO' = -90, 'OM' = -89, 'PK' = -88, 'PW' = -87, 'PS' = -86, 'PA' = -85, 'PG' = -84, 'PY' = -83, 'PE' = -82, 'PH' = -81, 'PN' = -80, 'PL' = -79, 'PT' = -78, 'PR' = -77, 'QA' = -76, 'RE' = -75, 'RO' = -74, 'RU' = -73, 'RW' = -72, 'BL' = -71, 'SH' = -70, 'KN' = -69, 'LC' = -68, 'MF' = -67, 'PM' = -66, 'VC' = -65, 'WS' = -64, 'SM' = -63, 'ST' = -62, 'SA' = -61, 'SN' = -60, 'RS' = -59, 'SC' = -58, 'SL' = -57, 'SG' = -56, 'SX' = -55, 'SK' = -54, 'SI' = -53, 'SB' = -52, 'SO' = -51, 'ZA' = -50, 'GS' = -49, 'SS' = -48, 'ES' = -47, 'LK' = -46, 'SD' = -45, 'SR' = -44, 'SJ' = -43, 'SZ' = -42, 'SE' = -41, 'CH' = -40, 'SY' = -39, 'TW' = -38, 'TJ' = -37, 'TZ' = -36, 'TH' = -35, 'TL' = -34, 'TG' = -33, 'TK' = -32, 'TO' = -31, 'TT' = -30, 'TN' = -29, 'TR' = -28, 'TM' = -27, 'TC' = -26, 'TV' = -25, 'UG' = -24, 'UA' = -23, 'AE' = -22, 'GB' = -21, 'UM' = -20, 'US' = -19, 'UY' = -18, 'UZ' = -17, 'VU' = -16, 'VE' = -15, 'VN' = -14, 'VG' = -13, 'VI' = -12, 'WF' = -11, 'EH' = -10, 'YE' = -9, 'ZM' = -8, 'ZW' = -7, 'OTHER' = 0, 'AF' = 1, 'AX' = 2, 'AL' = 3, 'DZ' = 4, 'AS' = 5, 'AD' = 6, 'AO' = 7, 'AI' = 8, 'AQ' = 9, 'AG' = 10, 'AR' = 11, 'AM' = 12, 'AW' = 13, 'AU' = 14, 'AT' = 15, 'AZ' = 16, 'BS' = 17, 'BH' = 18, 'BD' = 19, 'BB' = 20, 'BY' = 21, 'BE' = 22, 'BZ' = 23, 'BJ' = 24, 'BM' = 25, 'BT' = 26, 'BO' = 27, 'BQ' = 28, 'BA' = 29, 'BW' = 30, 'BV' = 31, 'BR' = 32, 'IO' = 33, 'BN' = 34, 'BG' = 35, 'BF' = 36, 'BI' = 37, 'CV' = 38, 'KH' = 39, 'CM' = 40, 'CA' = 41, 'KY' = 42, 'CF' = 43, 'TD' = 44, 'CL' = 45, 'CN' = 46, 'CX' = 47, 'CC' = 48, 'CO' = 49, 'KM' = 50, 'CD' = 51, 'CG' = 52, 'CK' = 53, 'CR' = 54, 'CI' = 55, 'HR' = 56, 'CU' = 57, 'CW' = 58, 'CY' = 59, 'CZ' = 60, 'DK' = 61, 'DJ' = 62, 'DM' = 63, 'DO' = 64, 'EC' = 65, 'EG' = 66, 'SV' = 67, 'GQ' = 68, 'ER' = 69, 'EE' = 70, 'ET' = 71, 'FK' = 72, 'FO' = 73, 'FJ' = 74, 'FI' = 75, 'FR' = 76, 'GF' = 77, 'PF' = 78, 'TF' = 79, 'GA' = 80, 'GM' = 81, 'GE' = 82, 'DE' = 83, 'GH' = 84, 'GI' = 85, 'GR' = 86, 'GL' = 87, 'GD' = 88, 'GP' = 89, 'GU' = 90, 'GT' = 91, 'GG' = 92, 'GN' = 93, 'GW' = 94, 'GY' = 95, 'HT' = 96, 'HM' = 97, 'VA' = 98, 'HN' = 99, 'HK' = 100, 'HU' = 101, 'IS' = 102, 'IN' = 103, 'ID' = 104, 'IR' = 105, 'IQ' = 106, 'IE' = 107, 'IM' = 108, 'IL' = 109, 'IT' = 110, 'JM' = 111, 'JP' = 112, 'JE' = 113, 'JO' = 114, 'KZ' = 115, 'KE' = 116, 'KI' = 117, 'KP' = 118, 'KR' = 119, 'KW' = 120, 'KG' = 121, 'LA' = 122, 'LV' = 123, 'LB' = 124, 'LS' = 125, 'LR' = 126, 'LY' = 127)
)
ENGINE = MergeTree
ORDER BY tuple()
SETTINGS index_granularity = 8192;

INSERT INTO iso_3166_1_alpha_2 SELECT (rand(number) % 256) - 128 FROM numbers(200000000);

OPTIMIZE TABLE iso_3166_1_alpha_2 FINAL;
```

Without cache:

```sql
SELECT count()
FROM iso_3166_1_alpha_2
WHERE c NOT IN ('CU', 'BN', 'VI', 'US', 'AQ', 'AG', 'AR', 'AM', 'AW', 'AU', 'AT', 'AZ', 'BS', 'BH', 'BD', 'BB', 'BY', 'BE')
SETTINGS max_threads = 1

Query id: 247e3bb5-eaec-46e7-8cad-05212870d249

┌───count()─┐
│ 185940192 │
└───────────┘

1 row in set. Elapsed: 1.175 sec. Processed 200.00 million rows, 200.00 MB (170.26 million rows/s., 170.26 MB/s.)
Peak memory usage: 21.20 KiB.
```

With cache:

```sql
SELECT count()
FROM iso_3166_1_alpha_2
WHERE c NOT IN ('CU', 'BN', 'VI', 'US', 'AQ', 'AG', 'AR', 'AM', 'AW', 'AU', 'AT', 'AZ', 'BS', 'BH', 'BD', 'BB', 'BY', 'BE')

Query id: 048db24a-472a-4eb0-966b-2c174ac546d1

┌───count()─┐
│ 185940192 │
└───────────┘

1 row in set. Elapsed: 0.329 sec. Processed 200.00 million rows, 200.00 MB (607.31 million rows/s., 607.31 MB/s.)
Peak memory usage: 72.28 KiB.
```

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 476549bdf391318b1fb6ebde1f579614f02cc97e with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55712/476549bdf391318b1fb6ebde1f579614f02cc97e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6545283308">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55712/476549bdf391318b1fb6ebde1f579614f02cc97e/ci_running.html">❌ failure</a></td></tr>
<tbody>
</table>

File: src/Interpreters/castColumn.cpp [hunk 1]
```diff
@@ -36,19 +41,27 @@ static ColumnPtr castColumn(const ColumnWithTypeAndName & arg, const DataTypePtr
...
 }
 
-ColumnPtr castColumnAccurateOrNull(const ColumnWithTypeAndName & arg, const DataTypePtr & type)
+FunctionBasePtr InternalCastFunctionCache::getOrSet(CastType cast_type, const String & from, const String & to, Getter && getter)
 {
-    return castColumn<CastType::accurateOrNull>(arg, type);
+    std::lock_guard lock{mutex};
+    auto key = std::make_tuple(cast_type, from, to);
+    auto it = impl.find(key);
+    return it != impl.end() ? it->second : impl.emplace(std::forward_as_tuple(cast_type, from, to), getter()).first->second;
```
Comment (nickitat) [inline src/Interpreters/castColumn.cpp:64]: maybe

```suggestion
    if (it == impl.end())
        it = impl.emplace(key, getter()).first;
    return it->second;
```

?

File: src/Interpreters/castColumn.h [hunk 2]
```diff
@@ -1,12 +1,22 @@
...
 
-ColumnPtr castColumn(const ColumnWithTypeAndName & arg, const DataTypePtr & type);
-ColumnPtr castColumnAccurate(const ColumnWithTypeAndName & arg, const DataTypePtr & type);
-ColumnPtr castColumnAccurateOrNull(const ColumnWithTypeAndName & arg, const DataTypePtr & type);
+struct InternalCastFunctionCache
+{
+    /// Maps <cast_type, from_type, to_type> -> cast functions
+    std::map<std::tuple<CastType, std::string_view, std::string_view>, FunctionBasePtr> impl;
+    mutable std::mutex mutex;
+    using Getter = std::function<FunctionBasePtr()>;
```
Comment (nickitat) [inline src/Interpreters/castColumn.h:14]: I'd prefer template parameter over conversion to `std::function`

File: src/Interpreters/castColumn.h [hunk 3]
```diff
@@ -1,12 +1,22 @@
...
 
 namespace DB
 {
 
-ColumnPtr castColumn(const ColumnWithTypeAndName & arg, const DataTypePtr & type);
-ColumnPtr castColumnAccurate(const ColumnWithTypeAndName & arg, const DataTypePtr & type);
-ColumnPtr castColumnAccurateOrNull(const ColumnWithTypeAndName & arg, const DataTypePtr & type);
+struct InternalCastFunctionCache
+{
+    /// Maps <cast_type, from_type, to_type> -> cast functions
```
Comment (nickitat) [inline src/Interpreters/castColumn.h:11]: these field could be declared `private`

**Review by nickitat — APPROVED** [review]

**Comment by canhld94** [comment]
There's data race, I will investigate tomorrow

**Comment by canhld94** [comment]
Integration tests failed but not related (S3, failover)

**Comment by nickitat** [comment]
`Integration tests (release) [2/4]` - #52554

**Pull request merged**